packages/agents/src/ai-chat-agent.ts

```diff
@@ -74,21 +82,31 @@ export class AIChatAgent<Env = unknown, State = unknown> extends Agent<
           [connection.id]
         );
         await this.#persistMessages(messages, [connection.id]);
+
+        const chatMessageId = data.id;
+        const abortSignal = this.#getAbortSignal(chatMessageId);
+
         return this.#tryCatch(async () => {
-          const response = await this.onChatMessage(async ({ response }) => {
-            const finalMessages = appendResponseMessages({
-              messages,
-              responseMessages: response.messages,
-            });
-
-            await this.#persistMessages(finalMessages, [connection.id]);
-          });
+          const response = await this.onChatMessage(
+            async ({ response }) => {
+              const finalMessages = appendResponseMessages({
+                messages,
+                responseMessages: response.messages,
+              });
+
+              await this.#persistMessages(finalMessages, [connection.id]);
+              this.#removeAbortController(chatMessageId);
+            },
+            abortSignal ? { abortSignal } : undefined
+          );
+
           if (response) {
             await this.#reply(data.id, response);
           }
         });
       }
       if (data.type === "cf_agent_chat_clear") {
+        this.#destroyAbortControllers();
```

Comment:
Note: Here, when you clear the chat, you'd also end up cancelling pending generations. Not sure if that's expected behavior